test/integration/replicaset/replicaset_test.go

```diff
@@ -515,3 +523,668 @@ func TestRSSelectorImmutability(t *testing.T) {
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-overlapping-rss", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+
+	// Create 2 RS with identical selectors
+	labelMap := map[string]string{"foo": "bar"}
+	for i := 0; i < 2; i++ {
+		// One RS has 1 replica, and another has 2 replicas
+		rs := newRS(fmt.Sprintf("rs-%d", i+1), ns.Name, i+1)
+		rs.Spec.Selector.MatchLabels = labelMap
+		rs.Spec.Template.Labels = labelMap
+		createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+		waitRSStable(t, c, rs, ns.Name)
+	}
+
+	// Expect 3 total Pods to be created
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := c.Core().Pods(ns.Name).List(options)
+	if err != nil {
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	if len(pods.Items) != 3 {
+		t.Errorf("len(pods) = %v, want %v", len(pods.Items), 3)
+	}
+
+	// Expect both RS have status.replicas = spec.replicas
+	for i := 0; i < 2; i++ {
+		replicaset, err := c.ExtensionsV1beta1().ReplicaSets(ns.Name).Get(fmt.Sprintf("rs-%d", i+1), metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("failed to get replicaset rs-%d: %v", i+1, err)
+		}
+		if replicaset.Status.Replicas != *replicaset.Spec.Replicas {
+			t.Fatalf(".status.replicas %d is not equal to .spec.replicas %d", replicaset.Status.Replicas, *replicaset.Spec.Replicas)
+		}
+	}
+}
+
+func TestPodOrphaningAndAdoptionWhenLabelsChange(t *testing.T) {
+	s, closeFn, rm, informers, c := rmSetup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("test-replicaset-basics", s, t)
+	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
+	rs := newRS("rs", ns.Name, 1)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go rm.Run(5, stopCh)
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	// Orphaning: RS should remove OwnerReference from a pod when the pod's labels change to not match its labels
+	// Although we can access the pod directly, we should still access it via selector to ensure the selector is working
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := c.Core().Pods(ns.Name).List(options)
+	if err != nil {
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
+	}
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	if len(pods.Items) != 1 {
+		t.Fatalf("len(pods) = %v, want %v", len(pods.Items), 1)
+	}
+
+	pod := &pods.Items[0]
```

Comment:
Now that you've fetched the Pod, check that `GetControllerOf(pod) != nil`. This proves that your later check for `GetControllerOf(pod) == nil` is meaningful.

Reply:
Good point! :)